monty-31/src/x86_64_avx512/poseidon2.rs

```diff
@@ -1,34 +1,438 @@
+
+                s0 -> (s0 + rc)^d
+                s -> (1 + Diag(v))s
+
+                Note that this matrix multiplication can be implemented as:
+                sum = sum_i s_i
+                s_i -> sum + s_iv_i
+
+                which is essentially how we implement it.
+            */
+
+            self.packed_internal_constants.iter().for_each(|&rc| {
+                state.s0 = add_rc_and_sbox::<FP, D>(state.s0, rc); // s0 -> (s0 + rc)^D
+                let sum_non_0 = sum_23(&transmute::<
+                    [__m512i; 23],
+                    [PackedMontyField31AVX512<FP>; 23],
+                >(state.s_hi)); // Get the sum of all elements other than s0.
+                ILP::diagonal_mul(&mut state.s_hi); // si -> vi * si for all i > 0.
+                let sum = sum_non_0 + state.s0; // Get the full sum.
+                state.s0 = sum_non_0 - state.s0; // s0 -> sum - 2*s0 = sum_non_0 - s0.
+                ILP::add_sum(
+                    &mut state.s_hi,
+                    transmute::<PackedMontyField31AVX512<FP>, __m512i>(sum),
+                ); // si -> si + sum for all i > 0.
+            })
+        }
+    }
+}
+
+/// Compute a collection of Poseidon2 external layers.
+/// One layer for every constant supplied.
+#[inline]
+fn external_rounds<FP, const WIDTH: usize, const D: u64>(
+    state: &mut [PackedMontyField31AVX512<FP>; WIDTH],
+    packed_external_constants: &[[__m512i; WIDTH]],
+) where
+    FP: FieldParameters,
+{
+    /*
+        The external layer consists of the following 2 operations:
+
+        s -> s + rc
+        s -> s^d
+        s -> Ms
+
+        Where by s^d we mean to apply this power function element wise.
+
+        Multiplication by M is implemented efficiently in p3_poseidon2/matrix.
+    */
+    packed_external_constants.iter().for_each(|round_consts| {
+        state
+            .iter_mut()
+            .zip(round_consts.iter())
+            .for_each(|(val, &rc)| {
+                *val = add_rc_and_sbox::<FP, D>(*val, rc);
+            });
+        mds_light_permutation(state, &MDSMat4);
+    });
+}
+
+impl<FP, const D: u64> ExternalLayer<PackedMontyField31AVX512<FP>, 16, D>
+    for Poseidon2ExternalLayerMonty31<FP, 16>
+where
+    FP: FieldParameters,
+{
+    type InternalState = InternalLayer16<FP>;
+
+    /// Compute the first half of the Poseidon2 external layers.
+    fn permute_state_initial(
+        &self,
+        mut state: [PackedMontyField31AVX512<FP>; 16],
+    ) -> Self::InternalState {
+        mds_light_permutation(&mut state, &MDSMat4);
+
+        external_rounds::<FP, 16, D>(&mut state, &self.packed_initial_external_constants);
+
+        InternalLayer16::from_packed_field_array(state)
+    }
+
+    /// Compute the second half of the Poseidon2 external layers.
+    /// SAFETY: The caller must ensure that each element of `state` represents a valid `MontyField31<PMP>`.
+    /// In particular, each element of each vector must be in `[0, P)` (canonical form).
```

Comment:
I'm a bit confused by this comment about safety. Should this function be marked as `unsafe`? It looks like the function is safe assuming `state` was initialised correctly, so maybe the `InternalLayer16` initialisation should be unsafe and the comment moved there?

Reply:
Hmm, marking this function as unsafe is a little troublesome as that would propagate through the trait system but maybe that's a worthwhile change?

Maybe I'll delete the outer comment but leave the inner one justifying why `to_packed_field_array` is safe to use. The function is still a little unsafe as if `InternalLayer16` doesn't contain canonical field elements things can break but this should never happen unless someone is doing something particularly strange.